File: library/src/main/java/com/github/ykrank/androidtools/util/L.java
package com.github.ykrank.androidtools.util;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.github.ykrank.androidtools.BuildConfig;
import com.github.ykrank.androidtools.GlobalData;
import com.orhanobut.logger.AndroidLogAdapter;
import com.orhanobut.logger.FormatStrategy;
import com.orhanobut.logger.Logger;
import com.orhanobut.logger.PrettyFormatStrategy;
import com.tencent.bugly.crashreport.BuglyLog;
import com.tencent.bugly.crashreport.CrashReport;

import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Created by AdminYkrank on 2016/4/20.
 * 对Log的包装
 */
public class L {
    static AtomicBoolean init = new AtomicBoolean(false);
    static boolean showLog = false;
    private static String TAG = BuildConfig.LIBRARY_PACKAGE_NAME;

    public static void init(@NonNull Context context) {
        TAG = GlobalData.provider.getLogTag();

        FormatStrategy formatStrategy = PrettyFormatStrategy.newBuilder()
                .showThreadInfo(true)  // (Optional) Whether to show thread info or not. Default true
                .methodCount(0)         // (Optional) How many method line to show. Default 2
                .methodOffset(7)        // (Optional) Hides internal method calls up to offset. Default 5
                //.logStrategy(customLog) // (Optional) Changes the log strategy to print out. Default LogCat
                .tag(TAG)   // (Optional) Global tag for every log. Default PRETTY_LOGGER
                .build();

        Logger.addLogAdapter(new AndroidLogAdapter(formatStrategy) {
            @Override
            public boolean isLoggable(int priority, String tag) {
                return showLog();
            }
        });
    }

    public static void setUser(final String id, final String name) {
        CrashReport.setUserId("id:" + id + ",name:" + name);
    }

    public static boolean showLog() {
        if (!init.get()) {
            init.set(true);
            showLog = GlobalData.provider.getDebug() || "alpha".equals(GlobalData.provider.getBuildType());
        }
        return showLog;
    }

    public static void l(String msg) {
        if (showLog()) {
            BuglyLog.d(TAG, msg);
        }
    }

    public static void print(String msg) {
        if (showLog()) {
            Log.d(TAG, msg);
        }
    }

    public static void print(Throwable e) {
        if (showLog() && e != null) {
            e.printStackTrace();
        }
    }

    public static void d(@Nullable String msg) {
        Logger.d(msg);
    }

    public static void d(@NonNull Throwable e) {
        Logger.d(e.getMessage());
    }

    public static void d(String tag, @Nullable String msg) {
        Logger.t(tag).d(msg);
    }

    public static void i(@NonNull String msg) {
        Logger.i(msg);
    }

    public static void i(String tag, @NonNull String msg) {
        Logger.t(tag).i(msg);
    }

    public static void w(@NonNull String msg) {
        Logger.w(msg);
    }

    public static void w(String tag, @NonNull String msg) {
        Logger.t(tag).w(msg);
    }

    public static void w(Throwable e) {
        Logger.e(e, "");
    }

    public static void e(@NonNull String msg) {
        e(null, msg, null);
    }

    public static void e(Throwable e) {
        e(null, "error", e);
    }

    public static void e(String msg, Throwable tr) {
        e(null, msg, tr);
    }

    public static void e(String tag, @NonNull String msg) {
        e(tag, msg, null);
    }

    public static void e(String tag, @NonNull String msg, Throwable tr) {
        BuglyLog.e(TAG + tag, msg, tr);
        if (tr != null) {
            tr.printStackTrace();
        }
        if (showLog() && tr != null) {
            CrashReport.postCatchedException(tr);
        }
    }

    public static void report(Throwable tr) {
        report(tr, Log.WARN);
    }

    public static void report(Throwable tr, int severity) {
        if (showLog()) {
            tr.printStackTrace();
        }

        ErrorParser errorParser = GlobalData.provider.getErrorParser();
        if (errorParser != null && errorParser.ignoreError(tr)) {
            return;
        } else {
            CrashReport.postCatchedException(tr);
        }
    }

    public static void report(String msg, Throwable tr) {
        leaveMsg(msg);
        report(tr);
    }

    public static void leaveMsg(String msg) {
        leaveMsg("MSG", msg);
    }

    public static void leaveMsg(String tag, String msg) {
        i(tag, msg);
        BuglyLog.i(tag, msg);
    }

    public static void leaveMsg(Throwable tr) {
        BuglyLog.e("MSG", "Error", tr);
    }

    public static void test() {
        throw new RuntimeException("Just test");
    }

    public static void throwNewErrorIfDebug(RuntimeException throwable) {
        if (BuildConfig.DEBUG) {
            throw throwable;
        } else {
            L.report(throwable, Log.WARN);
        }
    }
}

File: library/src/main/java/com/github/ykrank/androidtools/util/L.kt
package com.github.ykrank.androidtools.util

import android.content.Context
import android.util.Log
import com.github.ykrank.androidtools.BuildConfig
import com.github.ykrank.androidtools.GlobalData
import com.orhanobut.logger.AndroidLogAdapter
import com.orhanobut.logger.FormatStrategy
import com.orhanobut.logger.Logger
import com.orhanobut.logger.PrettyFormatStrategy
import com.tencent.bugly.crashreport.BuglyLog
import com.tencent.bugly.crashreport.CrashReport
import kotlinx.coroutines.CoroutineExceptionHandler
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Created by AdminYkrank on 2016/4/20.
 * 对Log的包装
 */
object L {
    var init = AtomicBoolean(false)
    var showLog = false
    private var TAG = BuildConfig.LIBRARY_PACKAGE_NAME

    val print = CoroutineExceptionHandler { _, e ->
        print(e)
    }

    fun init(context: Context) {
        TAG = GlobalData.provider.logTag
        val formatStrategy: FormatStrategy = PrettyFormatStrategy.newBuilder()
            .showThreadInfo(true) // (Optional) Whether to show thread info or not. Default true
            .methodCount(0) // (Optional) How many method line to show. Default 2
            .methodOffset(7) // (Optional) Hides internal method calls up to offset. Default 5
            //.logStrategy(customLog) // (Optional) Changes the log strategy to print out. Default LogCat
            .tag(TAG) // (Optional) Global tag for every log. Default PRETTY_LOGGER
            .build()
        Logger.addLogAdapter(object : AndroidLogAdapter(formatStrategy) {
            override fun isLoggable(priority: Int, tag: String?): Boolean {
                return showLog()
            }
        })
    }

    fun setUser(id: String?, name: String?) {
        CrashReport.setUserId("id:$id,name:$name")
    }

    fun showLog(): Boolean {
        if (!init.get()) {
            init.set(true)
            showLog = GlobalData.provider.debug || "alpha" == GlobalData.provider.buildType
        }
        return showLog
    }

    fun l(msg: String?) {
        if (showLog()) {
            BuglyLog.d(TAG, msg)
        }
    }

    fun print(msg: String?) {
        if (showLog()) {
            Log.d(TAG, msg!!)
        }
    }

    @JvmStatic
    fun print(e: Throwable?) {
        if (showLog() && e != null) {
            Log.i(TAG, "", e)
        }
    }

    @JvmStatic
    fun d(msg: String?) {
        Logger.d(msg)
    }

    fun d(e: Throwable) {
        Logger.d(e.message)
    }

    fun d(tag: String?, msg: String?) {
        Logger.t(tag).d(msg)
    }

    fun i(msg: String) {
        Logger.i(msg)
    }

    fun i(tag: String?, msg: String) {
        Logger.t(tag).i(msg)
    }

    fun w(msg: String) {
        Logger.w(msg)
    }

    fun w(tag: String?, msg: String) {
        Logger.t(tag).w(msg)
    }

    fun w(e: Throwable?) {
        Logger.e(e, "")
    }

    @JvmStatic
    fun e(msg: String) {
        e(null, msg, null)
    }

    @JvmStatic
    fun e(e: Throwable?) {
        e(null, "error", e)
    }

    @JvmStatic
    fun e(msg: String, tr: Throwable?) {
        e(null, msg, tr)
    }

    @JvmOverloads
    @JvmStatic
    fun e(tag: String?, msg: String, tr: Throwable? = null) {
        BuglyLog.e(TAG + tag, msg, tr)
        if (tr != null) {
            Log.e(TAG + tag, msg, tr)
        }
        if (showLog() && tr != null) {
            CrashReport.postCatchedException(tr)
        }
    }

    @JvmOverloads
    @JvmStatic
    fun report(tr: Throwable?, severity: Int = Log.WARN) {
        if (showLog()) {
            print(tr)
        }
        val errorParser = GlobalData.provider.errorParser
        if (errorParser != null && errorParser.ignoreError(tr!!)) {
            return
        } else {
            CrashReport.postCatchedException(tr)
        }
    }

    @JvmStatic
    fun report(msg: String, tr: Throwable?) {
        leaveMsg(msg)
        report(tr)
    }

    @JvmStatic
    fun leaveMsg(msg: String) {
        leaveMsg("MSG", msg)
    }

    fun leaveMsg(tag: String?, msg: String) {
        i(tag, msg)
        BuglyLog.i(tag, msg)
    }

    fun leaveMsg(tr: Throwable?) {
        BuglyLog.e("MSG", "Error", tr)
    }

    fun test() {
        throw RuntimeException("Just test")
    }

    fun throwNewErrorIfDebug(throwable: RuntimeException?) {
        if (BuildConfig.DEBUG) {
            throw throwable!!
        } else {
            report(throwable, Log.WARN)
        }
    }
}
